src/main/java/vazkii/botania/common/item/rod/ItemTornadoRod.java

```diff
@@ -142,27 +152,101 @@ public void onAvatarUpdate(IAvatarTile tile, ItemStack stack) {
+							cooldownTag.putInt(p.getUniqueID().toString(), 20);
+						}
+					} else if (p.getMotion().getY() > 0.3 && p.getMotion().getY() < 2 && !p.isElytraFlying()) {
+						doAvatarJump(p, world);
+						if (!world.isRemote) {
+							doAvatarMiscEffects(p, tile);
 						}
-					}
-
-					if (!world.isRemote) {
-						p.world.playSound(null, p.getPosX(), p.getPosY(), p.getPosZ(), ModSounds.dash, SoundCategory.PLAYERS, 1F, 1F);
-						p.addPotionEffect(new EffectInstance(ModPotions.featherfeet, 100, 0));
-						tile.receiveMana(-COST);
 					}
 				}
 			}
+			if (!world.isRemote) {
+				setAvatarCooldowns(te, cooldownTag);
+			}
+		}
+	}
+
+	private void doAvatarElytraBoost(PlayerEntity p, World world) {
+		Vector3d lookDir = p.getLookVec();
+		double mult = 1.25 * Math.pow(2.71828, -0.5 * p.getMotion().length());
+		p.setMotion(p.getMotion().getX() + lookDir.getX() * mult,
+				p.getMotion().getY() + lookDir.getY() * mult,
+				p.getMotion().getZ() + lookDir.getZ() * mult);
+
+		for (int i = 0; i < 20; i++) {
+			for (int j = 0; j < 5; j++) {
+				WispParticleData data = WispParticleData.wisp(0.35F + (float) Math.random() * 0.1F, 0.25F, 0.25F, 0.25F);
+				world.addParticle(data, p.getPosX() + lookDir.getX() * i,
+						p.getPosY() + lookDir.getY() * i,
+						p.getPosZ() + lookDir.getZ() * i,
+						0.2F * (float) (Math.random() - 0.5) * (Math.abs(lookDir.getY()) + Math.abs(lookDir.getZ())) + -0.01F * (float) Math.random() * lookDir.getX(),
+						0.2F * (float) (Math.random() - 0.5) * (Math.abs(lookDir.getX()) + Math.abs(lookDir.getZ())) + -0.01F * (float) Math.random() * lookDir.getY(),
+						0.2F * (float) (Math.random() - 0.5) * (Math.abs(lookDir.getY()) + Math.abs(lookDir.getX())) + -0.01F * (float) Math.random() * lookDir.getZ());
+			}
+		}
+	}
+
+	private void doAvatarJump(PlayerEntity p, World world) {
+		p.setMotion(p.getMotion().getX(), 2.8, p.getMotion().getZ());
+
+		for (int i = 0; i < 20; i++) {
+			for (int j = 0; j < 5; j++) {
+				WispParticleData data = WispParticleData.wisp(0.35F + (float) Math.random() * 0.1F, 0.25F, 0.25F, 0.25F);
+				world.addParticle(data, p.getPosX(),
+						p.getPosY() + i, p.getPosZ(),
+						0.2F * (float) (Math.random() - 0.5),
+						-0.01F * (float) Math.random(),
+						0.2F * (float) (Math.random() - 0.5));
+			}
+		}
+	}
+
+	private void doAvatarMiscEffects(PlayerEntity p, IAvatarTile tile) {
+		p.world.playSound(null, p.getPosX(), p.getPosY(), p.getPosZ(), ModSounds.dash, SoundCategory.PLAYERS, 1F, 1F);
+		p.addPotionEffect(new EffectInstance(ModPotions.featherfeet, 100, 0));
+		tile.receiveMana(-COST);
+	}
+
+	private CompoundNBT getAvatarCooldowns(TileEntity te) {
+		if (te.getTileData().contains(TAG_COOLDOWNS)) {
+			return ((CompoundNBT) te.getTileData().get(TAG_COOLDOWNS));
+		} else {
+			return new CompoundNBT();
+		}
+	}
+
+	private void setAvatarCooldowns(TileEntity te, CompoundNBT cooldownTag) {
+		te.getTileData().put(TAG_COOLDOWNS, cooldownTag);
```

Comment:
sorry for getting back to this late. can you not use getTileData and just call methods on TileAvatar directly? getTileData is not portable to Fabric.